File: Data Structures & Algorithms/balanced-binary-tree/submission-7.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        self.isBalanced = True

        def heightOf(node):
            if not node:
                return 0

            l = heightOf(node.left)
            r = heightOf(node.right)
            
            # otherwise
            b = abs(l - r) < 2
            self.isBalanced = self.isBalanced and b

            return max(l, r) + 1

        heightOf(root)
        return self.isBalanced

File: Data Structures & Algorithms/binary-tree-diameter/submission-7.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    int maxDiam = 0;

    int heightOf(TreeNode node) {
        if (node == null) return 0;

        int l = heightOf(node.left);
        int r = heightOf(node.right);

        this.maxDiam = Math.max(l + r, this.maxDiam);

        return Math.max(l, r) + 1;
    }

    public int diameterOfBinaryTree(TreeNode root) {
        if (root == null) return 0;
        this.heightOf(root);
        return this.maxDiam;
    }
}
